modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/metadata/IgniteMdSelectivity.java

```diff
@@ -17,49 +17,226 @@
+
+        for (RexNode node : operandsToProcess) {
+            RexNode res = getLocalRef(node);
+
+            // It can be null for case with expression for example: LENGTH('abc') = 3
+            if (res != null) {
+                List<SqlKind> vals = processOperands.computeIfAbsent(res, (k) -> new ArrayList<>());
+
+                vals.add(node.getKind());
+            } else {
+                baseSelectivity = Math.max(baseSelectivity, nonColumnRefSelectivity(call));
+            }
+        }
+
+        return computeOpsSelectivity(processOperands, baseSelectivity);
+    }
+
+    private static double nonColumnRefSelectivity(RexCall call) {
+        if (call.isA(SqlKind.EQUALS)) {
+            return EQ_SELECTIVITY;
+        } else if (call.isA(SqlKind.COMPARISON)) {
+            return 0.5;
+        } else {
+            // different OTHER_FUNCTION and other uncovered cases
+            return 0;
+        }
+    }
+
+    private static RexLocalRef getLocalRef(RexNode node) {
+        RexVisitor<Void> v = new RexVisitorImpl<>(true) {
+            @Override
+            public Void visitLocalRef(RexLocalRef locRef) {
+                throw new Util.FoundOne(locRef);
+            }
+        };
+
+        try {
+            node.accept(v);
+
+            return null;
+        } catch (Util.FoundOne e) {
+            return (RexLocalRef) e.getNode();
+        }
+    }
+
+    private static double guessSelectivity(@Nullable RexNode predicate) {
+        double sel = 1.0;
+        if ((predicate == null) || predicate.isAlwaysTrue()) {
+            return sel;
+        }
+
+        double artificialSel = 1.0;
+
+        // make set of AND ... AND calls
+        List<RexNode> conjunctions = RelOptUtil.conjunctions(predicate);
+
+        for (RexNode pred : conjunctions) {
+            // Expand sarg`s
+            RexNode predicateExpanded = expandSearch(Commons.rexBuilder(), null, pred);
+
+            if (predicateExpanded.isA(SqlKind.OR)) {
+                double processed = computeSelectivity((RexCall) predicateExpanded);
+                sel *= processed;
+            } else if (predicateExpanded.getKind() == SqlKind.IS_NOT_NULL) {
+                sel *= 0.9;
+            } else if (
+                    (predicateExpanded instanceof RexCall)
+                            && (((RexCall) predicateExpanded).getOperator()
+                            == RelMdUtil.ARTIFICIAL_SELECTIVITY_FUNC)) {
+                artificialSel *= RelMdUtil.getSelectivityValue(predicateExpanded);
+            } else if (predicateExpanded.isA(SqlKind.EQUALS)) {
+                sel *= 0.15;
+            } else if (predicateExpanded.isA(SqlKind.COMPARISON)) {
+                sel *= 0.5;
+            } else {
+                sel *= 0.25;
+            }
+        }
+
+        return sel * artificialSel;
+    }
+
     /**
      * GetSelectivity.
      * TODO Documentation https://issues.apache.org/jira/browse/IGNITE-15859
```

Comment:
we have a few documentation TODO in the class - will be great to fix it.

Reply:
done